benchmark/sample_genes.py

```diff
@@ -0,0 +1,90 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2015--, The WGS-HGT Development Team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+#
+# Select gene families (orthologous groups) from OrthoFinder results for
+# subsequent analyses
+#
+
+import sys
+import click
+import pandas as pd
+
+
+def sample_genes(ortho_groups_fp,
+                 min_taxa_cutoff=10.0):
+    """ Select gene families (orthologous groups) from OrthoFinder result
+
+    Parameters
+    ----------
+    ortho_groups_fp : str
+        orthologous groups definition file (OrthoFinder output file:
+        Orthogroups.csv)
+    min_taxa_cutoff : float
+        minimum number (if > 1) or fraction (if <= 1) of taxa that contain
+        one or more homologs of the query gene. default: 10.0
+
+    Return
+    ------
+    dict of dict of set of str
+        { ogid : { taxon : set(protein(s)) }
+            ogid: ID of the orthologous group (assigned by OrthoFinder)
+            taxon: ID of the taxon (genome)
+            protein: ID of the protein
+
+    Notes
+    -----
+    The purpose of this function is to pick gene families (orthologous groups,
+    or OGs) identified by OrthoFinder for subsequent analyses. Specifically,
+    the selected gene families will serve as input materials for phylogenetics-
+    based HGT detection methods. Because phylogenetic inference requires
+    adequate taxa to be statistically robust, OGs with too few members are
+    discarded, and HGTs (if any) in these gene families will be considered
+    unidentifiable due to inadequate taxon sampling.
+    """
+    ogs = pd.read_table(ortho_groups_fp, index_col=0, header=0, comment='#')
+    taxa = ogs.columns
+    # set minimum number of taxa cutoff
+    if min_taxa_cutoff > 1:
+        min_taxa_cutoff = int(min_taxa_cutoff)
+    elif 0 < min_taxa_cutoff <= 1:
+        min_taxa_cutoff = int(len(ogs.columns) * min_taxa_cutoff)
+    else:
+        raise ValueError('Error: %s is not a valid minimum taxa cutoff.'
+                         % min_taxa_cutoff)
+    genes = {}
```

Comment:
I am not sure if it's good for readability, but why not using more power of pandas:

```
x = ogs.applymap(lambda x: set(x.split(', ')) if pd.notnull(x) else {})
x[(x.applymap(len).sum(axis=1) >= min_taxa_cutoff) & (x['query'] != {})]
genes = x.T.to_dict()
```